### PR TITLE
Pass goal handle to goal response callback instead of a future

### PR DESCRIPTION
**Pull request by jacobperron**
This resolves an issue where `promise->set_value` is called before a potential call to `promise->set_exception`.
If there is an issue sending a result request, set the exception on the future to the result in the goal handle, instead of the future to the goal handle itself.

Alternative to https://github.com/ros2/rclcpp/pull/1292.

### CODE REVIEW

**Review by brawner — APPROVED** [review]
This seems reasonable to me. I'll let @hidmic have the final approval though.

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (hidmic) [inline rclcpp_action/include/rclcpp_action/client.hpp:646]: @jacobperron hmm, my only concern here is that the caller won't know that the result request went wrong till much later. It also means a caller can't cancel the goal just sent.

---

What if we just eat up the exception? It's not ideal, but `ClientGoalHandle::is_result_aware()` can still return `false` and `ClientGoalHandle::async_get_result()` (which will eventually be called anyways) can throw.

**Comment by brawner** [comment]
Here are the builds from #1290. I'm still trying to figure out the Windows issue (if it pops up)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12297/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7269/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10013/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12195/)

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp:646]: Perhaps it's not as ergonomic, but the caller can still immediately check the future returned by `Client::async_get_result`, and any error that occurred when sending the result request should be visible.

>  It also means a caller can't cancel the goal just sent.

This is a valid concern. I think what you're proposing is to remove the erasure logic below so that the user can still interact with the goal handle (even though it failed to get the result). I think calling `invalidate()` still makes sense so that calling `Client::async_get_result` still produces an exception.

Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp:646]: The other option I hadn't considered was to not handle the exception at all. This way we leave it up to the caller how to handle the exception (and they know something went wrong right away).

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp:645]: Correct me if I'm wrong, but I think this is the only type that can be thrown, [here](https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp_action/src/client.cpp#L334).

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (hidmic) [reply rclcpp_action/include/rclcpp_action/client.hpp:646]: Right. If `ClientGoalHandle::make_result_aware` fails, the goal handle will not be result aware. Any attempt to get the future [will result in an exception](https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L63-L65) that the user can handle.

Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp:647]: Take a look at 81239f1

As it is currently implemented, the user will get the [exception set in the future](https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L145) (because we set result awareness to `true` above, before sending a result request).

But, I think it makes sense to throw the error as earlier if possible, so I've make a small change make this happen: d444e09

Comment (hidmic) [reply rclcpp_action/include/rclcpp_action/client.hpp:647]: Do we have to do anything besides making the goal handle not result aware? Would it be wrong if the user attempts to request the result again?

**Comment by jacobperron** [comment]
Note to self: examples, demos, and other packages that register a result callback need to be updated.
I'll spend a little time looking at solutions for maintaining, but deprecated, the old signature.

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp:647]: I don't think there's anything else to do :thinking: If a user attempts to request the result I'm not actually sure what would happen, but I guess it shouldn't be allowed considering the result future would have already been set to an exception. In retrospect, I think it keeps the logic simpler to let the exception live in the future (1176dad).
Currently, the user is not able to request the result more than once (even if it is successful).

Comment (hidmic) [reply rclcpp_action/include/rclcpp_action/client.hpp:647]: Hmm, I see. We're not dealing with exceptions correctly e.g. depending on where `Client::async_get_result()` throws, the handle may be tagged as result aware (banning future result requests) or not. 

In my mind, a user that explicitly makes a goal handle result aware should get an exception if something goes wrong, and the goal handle should remain result unaware. That is, `Client::async_get_result()` should offer strong exception guarantees. It shouldn't be different for a user that implicitly makes a goal handle result aware by passing in a result callback to `Client::async_send_goal()`. If we don't make the latter throw (and there're valid arguments against), by keeping the goal handle result unaware we allow the user to check `ClientGoalHandle::is_result_aware()` or handle the exception that `ClientGoalHandle::async_get_result()` will throw. Does that make sense?

To invalidate the goal handle improves the situation quite a bit, but it delays error generation (until `future::get`, when it could show itself before spinning), it induces an spurious status change (to UNKNOWN until another status message comes along), and it prevents any retries on an otherwise unaware goal handle. 

But I won't block you. CI's green and this is a fairly unusual case :)

**Review by hidmic — APPROVED** [review]

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Yeah, that makes sense to me. I've made changes so that in the event that a result request fails, `ClientGoalHandle::is_result_aware()` should return `false` and `Client::async_get_result()` raises an exception (9cb1776).

I couldn't find a nice way to do it with the existing state information, so I've introduced a new member to `ClientGoalHandle` for checking if `invalidate()` was called, which doubles as an improved error message.

PTAL

**Comment by jacobperron** [comment]
Here's of list of downstream changes required by this change:

- https://github.com/ros2/demos/pull/463
- https://github.com/ros2/examples/pull/291
- https://github.com/ros2/geometry2/pull/323
- https://github.com/ros/ros_tutorials/pull/100

I couldn't find a nice way to deprecate the old signature. Suggestions welcome that would make the change easier on users.

**Review by hidmic — APPROVED** [review]

File: rclcpp_action/include/rclcpp_action/client.hpp
Comment (hidmic) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Neat! Two questions. Could a goal handle be invalidated twice? Should we do something about the spurious UNKNOWN status on invalidation?

Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Good questions. I think we certainly want to guard against multiple calls invalidate (see b145c73). As an aside, I wish it was easier to test this stuff :upside_down_face: It might be worth revisiting the architecture to make validation easier.

I'm not sure what to do about the UNKNOWN status. We could remove setting it when the goal is invalidated. Really I think the status should be reporting the latest state from the action server. Since this is keeping the same behavior (ie. when the client is destroyed), I'll defer deciding to remove the unknown status to another PR.

**Comment by jacobperron** [comment]
Testing all connected PRs:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12331/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7300/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10044/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12235/)

**Comment by brawner** [comment]
CI is all green, is this ready to be merged?

**Comment by jacobperron** [comment]
Err, I may have merged too early. The connected PRs need approvals too :fearful:

**Comment by ivanpauno** [comment]
New build failure in nightlies seem to be related with this: https://ci.ros2.org/view/nightly/job/nightly_linux_release/1681/.

**Comment by jacobperron** [comment]
Added a release note for Galactic: https://github.com/ros2/ros2_documentation/pull/871

**Comment by v-lopez** [comment]
I apologise for coming back to this merged changes, but have you found any way of keeping compatibility between new and old code? Or the only solution right now is a branch for foxy and new developments on a separate branch?

**Comment by jacobperron** [comment]
@v-lopez Unfortunately, the original bug was "easily" fixed by changing API. In this case, I think the API break makes sense since the original API of passing a future was not necessary and part of the bug (see discussion in https://github.com/ros2/rclcpp/pull/1292#discussion_r486579163). 

I think we can easily maintain backwards compatibility if we can use C++17, specifically `std::variant`. E.g. change the type of the goal callback member of the `SendGoalOptions` struct to a `std::variant`. I don't know if we've settled on the minimal C++ standard for Galactic. If we can use C++17, I'm happy to add a change to make this backward compatible.

**Comment by v-lopez** [comment]
In my use cases I've managed to workaround it using the returned shared_future from sending the goal.
My idea was more on the lines of adding the new callback to the foxy version somehow, since it is LTS a lot of packages are going to use it for the next years, and the code won't compile on galactic. But I cannot think of a way of doing it easily right now.

**Comment by jacobperron** [comment]
I don't think adding the new callback to Foxy can be done in an ABI (or API) compatible way. Also, my understanding is the error this bug causes happens rarely (if at all), so I think it's very unlikely that we would backport this change.

**Comment by jacobperron** [comment]
@v-lopez Please correct me if I'm wrong. If you are experiencing errors due to the bug this change fixes, perhaps we could patch it in Foxy a different way.

**Comment by v-lopez** [comment]
No, I've not experienced the bug at all.

Was just looking for a way to make the code compatible Foxy and onwards, luckily there's multiple ways of getting the GoalHandle, we can go with that. Thanks for your efforts.

**Comment by ivanpauno** [comment]
> I think we can easily maintain backwards compatibility if we can use C++17, specifically std::variant. E.g. change the type of the goal callback member of the SendGoalOptions struct to a std::variant. I don't know if we've settled on the minimal C++ standard for Galactic. If we can use C++17, I'm happy to add a change to make this backward compatible.

Though using `std::variant` is nice, for this case we can manually implement an API backwards compatible adapter:

```cpp
  struct SendGoalOptions
  {
    SendGoalOptions()
    : goal_response_callback(nullptr),
      feedback_callback(nullptr),
      result_callback(nullptr)
    {
    }

    /// Function called when the goal is accepted or rejected.
    /**
     * Takes a single argument that is a goal handle shared pointer.
     * If the goal is accepted, then the pointer points to a valid goal handle.
     * If the goal is rejected, then pointer has the value `nullptr`.
     */
    GoalResponseCallbackCompatibilitySelector goal_response_callback;

    /// Function called whenever feedback is received for the goal.
    FeedbackCallback feedback_callback;

    /// Function called when the result for the goal is received.
    ResultCallback result_callback;
  };

class GoalResponseCallbackCompatibilitySelector {
public:
  // implicit constructor
  GoalResponseCallbackCompatibilitySelector(GoalResponseCallbackOldSignature old_callback) : old_callback_(old_callback)
{}

  // implicit constructor
  GoalResponseCallbackCompatibilitySelector(GoalResponseCallbackNewSignature new_callback) : new_callback_(new_callback)
{}

private:
  friend rclcpp_action::Client;  // to avoid exposing getters of new_callback and old_callback publicly

  // Consumer of the class will first try to use new_callback_ and if it is `nullptr`, `get_old_callback` will be used.
  GoalResponseCallbackNewSignature new_callback_;
  GoalResponseCallbackOldSignature old_callback_;
};
```

Considering this change might affect several users, I think it's worth going through a ping-pong deprecation cycle.

**Comment by jacobperron** [comment]
Thanks for the suggestion, @ivanpauno  :bow: 
I agree, we avoid a hard break and go through a deprecation cycle. Would you like to open a PR applying your suggestion? Otherwise, I'm happy to do it.

**Comment by ivanpauno** [comment]
@v-lopez see #1495, which enabled backwards compatibility.

**Pull request merged**
